public/app/features/templating/state/sharedReducer.ts

```diff
@@ -0,0 +1,171 @@
+import { createReducer } from '@reduxjs/toolkit';
+import cloneDeep from 'lodash/cloneDeep';
+
+import {
+  addVariable,
+  changeVariableOrder,
+  changeVariableProp,
+  changeVariableType,
+  duplicateVariable,
+  removeInitLock,
+  removeVariable,
+  resolveInitLock,
+  setCurrentVariableValue,
+  storeNewVariable,
+} from './actions';
+import { VariableModel, VariableOption, VariableWithOptions } from '../variable';
+import { ALL_VARIABLE_VALUE, emptyUuid, getInstanceState } from './types';
+import { variableAdapters } from '../adapters';
+import { changeVariableNameSucceeded, variableEditorUnMounted } from '../editor/reducer';
+import { Deferred } from '../deferred';
+import { changeToEditorEditMode } from './uuidInEditorReducer';
+import { initialVariablesState } from './variablesReducer';
+import { isQuery } from '../guard';
+
+export const sharedReducer = createReducer(initialVariablesState, builder =>
+  builder
+    .addCase(addVariable, (state, action) => {
+      state[action.payload.uuid!] = cloneDeep(variableAdapters.get(action.payload.type).initialState);
+      state[action.payload.uuid!] = {
+        ...state[action.payload.uuid!],
+        ...action.payload.data.model,
+      };
+      state[action.payload.uuid!].uuid = action.payload.uuid;
+      state[action.payload.uuid!].index = action.payload.data.index;
+      state[action.payload.uuid!].global = action.payload.data.global;
+      state[action.payload.uuid!].initLock = new Deferred();
+    })
+    .addCase(resolveInitLock, (state, action) => {
+      const instanceState = getInstanceState(state, action.payload.uuid!);
+      instanceState.initLock?.resolve();
+    })
+    .addCase(removeInitLock, (state, action) => {
+      const instanceState = getInstanceState(state, action.payload.uuid!);
+      instanceState.initLock = null;
+    })
```

Comment:
Can't see where in the new code initLock is used. Is it necessary?

Reply:
Some magic code migrated from variable_srv in `public/app/features/templating/state/actions.ts`